Rollup counts for concept entities.

marikomedlock opened the pull request:
Added a new indexing workflow to precompute rollup counts for the concept entities: condition, procedure, ingredient, measurement, visit, observation, device. The rollup counts indicate how many people have at least one occurrence of the concept. e.g. How many people have at least one occurrence of "diabetes" (condition)/ "leg surgery" (procedure).

The workflow takes 3 inputs:
1. list of all concept entities (e.g. all conditions)
2. list of alloccurrence entities (e.g. all condition occurrences, only need the `concept_id` and `person_id` fields)
3. (only for hierarchical entities) list of all ancestor-descendant relationships (e.g. among conditions)

Generated the indexed tables for all `aou_synthetic` entities in the `broad-tanagra-dev` project. Updated the worklog with the commands run.

A follow-on PR will expose the rollup counts for each entity as a new entity attribute in the underlay, so that the UI can use this.

calbach reviewed (commented):
Nice, overall this was pretty easy to follow. I will need to re-review countDistinct after some doc clarifications.

marikomedlock commented on workflow/src/main/java/bio/terra/tanagra/workflow/CountUtils.java:
```diff
@@ -0,0 +1,160 @@
+   * Count the number of distinct auxiliary nodes that each primary node has.
+   *
+   * @param primaryNodes a collection of the primary nodes that we want a count for
+   * @param auxiliaryNodes a collection of all auxiliary nodes that we want to count
```
It's the pair of `(node, what_to_count)` e.g. `(condition, person)`. Expanded the documentation here a bit and added an example.

marikomedlock commented on workflow/src/main/java/bio/terra/tanagra/workflow/CountUtils.java:
```diff
@@ -0,0 +1,160 @@
+  private CountUtils() {}
+
+  /**
+   * Count the number of distinct auxiliary nodes that each primary node has.
```
Renamed auxiliary -> occurrence and added an example here. I struggled with how much of the primary/occurrence/concept (e.g. person/condition_occurrence/condition) relationships should be documented here, because this indexing workflow is really just doing a `count` with a `group by` clause. We may want to use this logic for other entity relationships in the future (e.g. which days did we diagnose someone with covid = count condition_occurrences grouped by the date). So I tried to name things here in relation to what the workflow is doing, as opposed to what types of entities and relationships we're using this for now. I see this workflow as having 3 parts: the nodes we want to generate counts for (called them "primary"), the occurrences that we want to process (called them "occurrences"), and the field that we want to count distinct occurrences of (called this "what_to_count"). Suggestions for better names are much appreciated.

calbach replied:
This seems reasonable for now. Probably needs some broader naming discussion to avoid conflicts before moving to something else. A few observations:

1. Occurrence may need some first class treatment in the underlay. Dealing with full generality here at the indexing layer would otherwise require some very vague language or generic and maybe non-performant functionality, e.g. you might describe this as "a hierarchical summing entity aggregator which takes an arbitrary counter expression". Or expanded for this specific case (still maintaining a degree of performance): a hierarchical aggregator on an entity1 by way of a specification relationship1 to unique instances of entity2.
2. The primary target relationship seems important here; and this code could perhaps be simplified if we have a term for this - i.e. in OMOP, the person table. It's a third variable which perhaps could act as more of a global concept.
3. The relationship of concept <- occurrence -> person seems important, and might want some special treatment.

marikomedlock replied:
I agree the person/concept/occurrence pattern needs first class treatment in the underlay. Tim and I are still discussing what this should look like, but the current requirement that users define essentially identical relationships for each version of this pattern (e.g. for conditions, procedures, devices, etc.) is something we want to do away with. We should have some concept of an occurrence group of entities, where the user just tells us which ones they are and we infer the relationships and indexing workflows needed.

I'm not sure yet how much of that logic should live in the workflow classes themselves, as opposed to the code that generates the list of workflow commands to run.

Also agreed on the need for some better names here to avoid conflicts. Maybe this should be part of the general underlay overhaul.

marikomedlock commented on workflow/src/main/java/bio/terra/tanagra/workflow/CountUtils.java:
```diff
@@ -0,0 +1,160 @@
+  }
+
+  /**
+   * Aggregate counts in a hierarchy, so that nodes have a count that includes all their
```
I removed this helper method, but fixed this wording in the documentation of the Beam workflow class.

marikomedlock commented on workflow/src/main/java/bio/terra/tanagra/workflow/CountUtils.java:
```diff
@@ -0,0 +1,160 @@
+   *
+   * @param nodeDirectCount a collection of (node, count) mappings where the count is the direct
+   *     count for that node only, not including descendants
+   * @param descendantAncestor a collection of all (descendant, ancestor) pairs
```
Fixed.

marikomedlock commented on workflow/src/main/java/bio/terra/tanagra/workflow/PrecomputeCounts.java:
```diff
@@ -0,0 +1,185 @@
+ * <p>Primary = Condition. We want to generate a count for each condition. Input is a list of
+ * conditions.
+ *
+ * <p>Secondary = Condition_Occurrence. We want to count the number of people that have at least one
```
Fixed and renamed auxiliary -> occurrence everywhere.

marikomedlock commented on workflow/src/test/java/bio/terra/tanagra/workflow/CountUtilsTest.java:
```diff
@@ -0,0 +1,192 @@
+  @Test
+  public void allNodesHaveNonZeroCount() {
+    Multimap<Long, Long> auxiliaryNodes = MultimapBuilder.hashKeys().arrayListValues().build();
+    auxiliaryNodes.putAll(10L, List.of(20L));
```
Agreed. I poked around a bit online, but didn't see a way to make this less verbose.

marikomedlock commented:
While doing more spot checking of the counts, I found a problem with my original workflow code for entities with a hierarchy. It was double counting people for a parent node when they have occurrences of two different child nodes. I've fixed this now and regenerated all the `broad-tanagra-dev` index tables with this updated workflow. I've also spot checked them against the source data and the AoU Cohort Builder test instance. The counts are not identical to the test instance, but look much closer than before for the higher-level nodes in the hierarchy.

Also added more documentation and examples for the workflow code.

marikomedlock commented on workflow/src/main/java/bio/terra/tanagra/workflow/BigQueryUtils.java:
```diff
@@ -33,21 +33,48 @@ public static PCollection<Long> readNodesFromBQ(
+  }
+
+  /**
+   * Read all the auxiliary rows from BQ and build a {@link PCollection} of {@link KV} pairs (node,
```
Fixed.

tjennison-google commented on workflow/src/main/java/bio/terra/tanagra/workflow/PrecomputeCounts.java:
```diff
@@ -0,0 +1,195 @@
+    Pipeline pipeline = Pipeline.create(options);
+
+    // read in the queries from files
+    String primaryNodesQuery = Files.readString(Path.of(options.getAllPrimaryNodesQuery()));
```
You could consider structuring this like I did for WriteAllNodes. That moves the queries and other configuration into the underlay rather than separate files so it's easy to run using RunUnderlayWorkflows, even if there's not actually any smarter logic there yet.

marikomedlock replied:
Yes, I'm planning to refactor all the Beam workflows together in a separate PR.

calbach commented on workflow/src/main/java/bio/terra/tanagra/workflow/CountUtils.java:
```diff
@@ -0,0 +1,192 @@
+  public static PCollection<KV<Long, Long>> countDistinct(
+      PCollection<Long> primaryNodes, PCollection<KV<Long, Long>> occurrences) {
+    // remove duplicate occurrences
+    PCollection<KV<Long, Long>> distinctOccurrences =
```
My understanding is that Distinct would make this unique by KV - my question was whether you could still have duplicate keys. I guess this is implicit in terms of the nature of this input collection though, that you wouldn't have any duplicate keys

calbach replied:
Sorry, I was confusing the value here as a count rather than the person ID. Ignore.

marikomedlock commented on workflow/src/main/java/bio/terra/tanagra/workflow/CountUtils.java:
```diff
@@ -0,0 +1,192 @@
+    // count the number of occurrences per primary node
+    // note that this will not include any zeros -- only primary nodes that have >=1 auxiliary node
+    // will show up here
+    PCollection<KV<Long, Long>> nonZeroCountKVs =
```
Changed it to primaryNodeInitialKVs and countKVs. Also added some more comments about what's happening in the ParDo through the JOIN result.

marikomedlock commented on workflow/src/main/java/bio/terra/tanagra/workflow/CountUtils.java:
```diff
@@ -0,0 +1,192 @@
+                  context.output(KV.of(node, nonZeroCountTagIter.next()));
+                } else {
+                  context.output(
+                      KV.of(node, zeroCountTagIter.next())); // equivalent to KV.of(node, 0L)
```
Changed to explicit 0L.
